File: part2/transports/memory/mem_trans.cpp
// Apache Thrift Memory Transport in C++ 

#include <thrift/transport/TBufferTransports.h>
#include <iostream>

struct Trade {
   char symbol[16];
   double price;
   int size;
};

int main() {
    apache::thrift::transport::TMemoryBuffer trans(4096);

    Trade t{"F", 13.10, 2500};
    trans.write(reinterpret_cast<const uint8_t *>(&t), sizeof(t));

    Trade tin;
    int n = trans.read(reinterpret_cast<uint8_t *>(&tin), sizeof(tin));

    std::cout << "Trade(" << n << "): " << tin.symbol << " "
              << tin.size << " @ " << tin.price << std::endl;
}

File: part2/transports/memory/mem_trans.py
# Apache Thrift Memory Transport in Python 

import pickle
from thrift.transport import TTransport

class Trade:
    def __init__(self, symbol="", price=0.0, size=0):
        self.symbol=symbol
        self.price=price
        self.size=size

trans = TTransport.TMemoryBuffer()
trade = Trade("F", 13.10, 2500)
trans.write(pickle.dumps(trade))

trans.cstringio_buf.seek(0)
bstr = trans.read(4096)
trade_read = pickle.loads(bstr)
print("Trade(%d): %s %d @ %f" % (len(bstr), trade_read.symbol, trade_read.size, trade_read.price))

